src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,56 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var dependency in dependencies)
+                    {
+                        var (curr, _) = dependency;
```

Comment:
A small suggestion regarding the variable name "curr". It is not vary clear what that variable represents, so maybe renaming it to "dependencyNode" might better convey what the variable represents.

Since the "dependency" word is used a lot in this code fragment, maybe the "dependency" variable could be named "entry"?

Reply:
In this case, we know that `curr` is related to a dependency because it was extracted from a variable called `dependency`, so I think that `dependencyNode` might be too long for a local variable name. Maybe something like `target` would work, since it is the target of a dependency edge in the graph.

Reply:
`curr`, as well as `prev` and `next` are commonly used identifiers in algorithms that walk through nodes of a data structure. I was going to rename it to `target` as per Sarah's suggestion, but after thinking about it, that seems less clear to me given this precedent.